xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -444,6 +555,177 @@ public void run() {
         });
       }
     }
+
+    @Nullable
+    private Long determineFaultDelayNanos(FaultDelay faultDelay, Metadata headers) {
+      Long delayNanos;
+      Integer delayRate = null;
+      if (faultDelay.headerDelay()) {
+        try {
+          int delayMillis = Integer.parseInt(headers.get(HEADER_DELAY_KEY));
+          delayNanos = TimeUnit.MILLISECONDS.toNanos(delayMillis);
+          String delayPercentageStr = headers.get(HEADER_DELAY_PERCENTAGE_KEY);
+          if (delayPercentageStr != null) {
+            int delayPercentage = Integer.parseInt(delayPercentageStr);
+            delayRate = delayPercentage * 100 * 100;
+          }
+        } catch (NumberFormatException e) {
+          return null; // treated as header_delay not applicable
+        }
+      } else {
+        delayNanos = faultDelay.delayNanos();
+      }
+      if (delayRate == null) {
+        delayRate = faultDelay.ratePerMillion();
+      }
+      if (random.nextInt(1_000_000) >= delayRate) {
+        return null;
+      }
+      return delayNanos;
+    }
+
+    @Nullable
+    private Status determineFaultAbortStatus(FaultAbort faultAbort, Metadata headers) {
+      Status abortStatus = null;
+      Integer abortRate = null;
+      if (faultAbort.headerAbort()) {
+        try {
+          String httpCodeStr = headers.get(HEADER_ABORT_HTTP_STATUS_KEY);
+          if (httpCodeStr != null) {
+            int httpCode = Integer.parseInt(httpCodeStr);
+            abortStatus = GrpcUtil.httpStatusToGrpcStatus(httpCode);
+          }
+          String grpcCodeStr = headers.get(HEADER_ABORT_GRPC_STATUS_KEY);
```

Comment:
According to Envoy:
> If both x-envoy-fault-abort-request and x-envoy-fault-abort-grpc-request headers are set then x-envoy-fault-abort-grpc-request header will be ignored and fault response http status code will be set to x-envoy-fault-abort-request header value.

Envoy prefers http status code over grpc status code. 

But we may want confirmation from Mark (previously we encountered similar case for `GrpcTimeoutHeaderMax` vs `MaxStreamDuration`).